### ISSUE TITLE
ah buffer is never release if request had content

### ISSUE THREAD

**Issue by rickynelson1939** [warmcat/libwebsockets#944]
ah buffer is never released if request has content.
tested on 2.2
The test I did:
`info.max_http_header_pool = 4;`
4 http connections do POST 1 byte on "test-server" every 4 seconds.
5th connection tries to do same request - it wait till timeout, debugging shows that this line: `pt->ah_count_in_use--; `
was never executed.

**Comment by lws-team** [comment]
It sounds like a bug somehere along the line... the connections that did the POST closed?  Or they are still open?

**Comment by rickynelson1939** [comment]
Connections are still open, and they should be open in order to hold ah buffers, however ah buffers are released after I close connections.

**Comment by lws-team** [comment]
So the ah are not leaking... they're just held for too long.

At the moment for ws we drop the ah just after ws upgrade is established.  For GET we drop it if we completed the transaction and there are no pipelined headers already waiting on the connection.

But for POST I am not sure what we do.  It could probably get more strict about dropping the ah for both GET and POST after giving the user code one chance to look at the headers and let the POST payload stuff complete without any ah; it's no longer dealing with headers at that point.

I'll look at it later and reply again.

**Comment by lws-team** [comment]
Please try the patch above (on master)... it drops the ah immediately after CALLBACK_HTTP.

In addition it makes all the ah access functions check for NULL and report an lwsl_err() clearly if they were used after the ah had gone.

**Comment by rickynelson1939** [comment]
I did the test, unfortunately a buffer is still released only after connection is closed. It may be because LWS_CALLBACK_HTTP event is firing not at line 1181, but at line 951 instead.

**Comment by lws-team** [comment]
Yeah... I was looking for where it handled POST, but that is not the only way to handle POST.

It looks like it has to be done everywhere, which might have some side effects on GET.  But I guess that bullet will have to be bitten... I'll try that here first and see what happens.

**Comment by rickynelson1939** [comment]
I hate to say this, but GET can contain content (body) as well and has the very same effect, so...
And I have an offtopic question, I cannot manage get to work POST via chunked transfer-encoding, is this supported by lws?

**Comment by lws-team** [comment]
It seems the reason it's like this atm is when there is still known to be payload on the socket, until it has parsed it, it doesn't know if the payload includes another pipelined set of http/1.1 headers afterwards (for which it wants the ah to still be around to parse again).  So a situation like

```
[hdrs for POST /form] [ body related to POST] [hdrs for GET /favicon.ico]
```

may all be sent in one packet.... this is a bit contrived with POST but I have seen it happen with GET.

Ideally GET should be strict about dropping the ah too, so that's more of an opportunity than a problem, but it pushed the problem to confirming when the transaction completes, we no longer always arrive at that point with a valid ah inherited from the first transaction, that it acts well.

AFAIK chunked isn't supported on POST... it supports what browsers emit, which is multipart... that has its own headers for sizing content.

**Comment by rickynelson1939** [comment]
This GET with content, I'm not sure if browser can send something like that, but some attacker sure can and the main concern about all this is that any request with content will hold ah until request without content is send or connection is closed, so enough amount of connections, that send post every n seconds to move timeout forward can render a "denied of service" situation, while any other connections are sent to wait list. This trick works with POST and GET, so I would not pay special attention to POST in particular. I'm saying this to make sure there is no misunderstanding in this whole thing.

**Comment by lws-team** [comment]
The connections should have timeouts, for connection phases lws controls they already do have, and you can control how long the connection can live in user code.  You can also declare more ah in the pool.

Before I introduced the ah pool concept, lws just allocated an ah when dealing with any inbound connection.  So you could equally well DoS the whole server by making a large number of allocations leading to an unbounded anount of heap allocation.  On a small platform that is fatal.

We are better off this way already... being strict about how long the ah is held is an optimization, timeouts should enforce availability generally.  This will be a nice improvement but it is not staving off the end of the world ^^

**Comment by rickynelson1939** [comment]
Is an assumption that ah buffer can be held before guaranteed reuse not more than timeout_secs seconds correct? The way I see it, ah buffer can be hold in post or pipeline loop for indefinite amount of time, so even if I set the timeout_secs to 1 second some connection will be able to hold ah for themselves so it's like they are allocating it only for themselves from the pool. And yeah, in this situation the only way to deal with this is either disallow any body in requests and pipelining, or allocate pool large enough so every connection can get it's individual ah which is defeating the purpose of having pool at all. Okay, it's your vision, sure, but I think the main advantage of having pool is to maintain reusability for ah and, hopefully, do it in a orderly fashion, so every connection could process its requests no matter if someone do greedy pipelining or posting something. I don't know, I guess this make sense :)

**Comment by lws-team** [comment]
The point is every wsi has a timeout, it has nothing to do with ah-es.  If you set the timeout active on the wsi, unless you do something to extend his timeout again, the connection will close and all resources freed.  (And unlike the apocolypitc title on this issue, we already confirmed there is no leaking, closing the connection is enough). If you care about endless connections, manage the timeout on the connection.

Thanks for taking the time to take a shit on the idea of ah-es.  As I wrote, having a waiting list for running out of a constrained resource is a lot more controlled than exhausting the global heap by allocations without any limit.

At any rate, I used enough of my Sunday on this.  If you have a fix, send a patch, this is not the most important thing to work on for me.
